src/databricks/labs/ucx/progress/tables.py

```diff
@@ -38,17 +42,23 @@ def __init__(
             schema,
             table,
         )
-        self._table_migration_index = table_migration_index
+        self._migration_status_refresher = migration_status_refresher
 
-    def _encode_record_as_historical(self, record: Table) -> Historical:
+    SC: ClassVar = TableMigrationIndex
+
+    @contextmanager
+    def _snapshot_context(self) -> Generator[SC, None, None]:
+        yield TableMigrationIndex(self._migration_status_refresher.snapshot())
+
+    def _encode_record_as_historical(self, record: Table, snapshot_context: SC) -> Historical:
```

Comment:
> Prior to these changes, this was being injected via the constructor and could lead to mismatches because it needs to correspond to the Table snapshot that is being encoded. 

why `force_refresh=True` is not enough there?

Reply:
> Drop `@final` on the base-class method, and simply specialise it in TableProgressEncoder to handle both loading the index and supplying it as a parameter for encoding.

this is the preferred way, as it the least amount of change.

> Update the signature of ProgressEncoder so that the caller is responsible for providing both the Table snapshot to encode, as well as the accompanying TableMigrationIndex that it should use for enrichment. This is also a public API change. However I don't really see it as an improvement, and it makes some things internally more complicated to deal with.

it doesn't seem to be the least amount of change. Also, other encoders don't need tables.

Reply:
Okay, I've ditched the context and switched to specialisation just within `TableProgressEncoder`.

Regarding `force_refresh=True`, the issue is not so much about whether it's enough: it's about where it occurs and which component is responsible for it. My understanding based on [this](https://github.com/databrickslabs/ucx/pull/2743#discussion_r1783326822) is that workflows are where force-refreshing is supposed to happen explicitly, and not implicitly within other components. The changes here are aimed at moving closer to that.

Reply:
I've extracted this part of the PR to #3270 to: a) reduce the size and scope of this PR; b) it's clearer what's going on.